libretro-common/include/lists/linked_list.h

```diff
@@ -0,0 +1,298 @@
+ * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#ifndef __LIBRETRO_SDK_LINKED_LIST_H
+#define __LIBRETRO_SDK_LINKED_LIST_H
+
+#include <retro_common_api.h>
+
+#include <boolean.h>
+#include <stddef.h>
+
+RETRO_BEGIN_DECLS
+
+/**
+ * Represents a linked list. Contains any number of elements.
+ */
+typedef struct linked_list linked_list_t;
+
+/**
+ * Represents an iterator for iterating over a linked list. The iterator can
+ * go through the linked list forwards or backwards.
+ */
+typedef struct linked_list_iterator linked_list_iterator_t;
+
+/**
+ * Creates a new linked list with no elements.
+ *
+ * @return New linked list
+ */
+linked_list_t *linked_list_new();
+
+/**
+ * @brief frees the memory used by the linked list
+ * 
+ * Frees all of the memory used by this linked list. The values of all
+ * remaining elements are freed using the "free_value" function. Does
+ * nothing if "list" is NULL.
+ *
+ * @param list linked list to free
+ * @param free_value function to use to free remaining values
+ */
+void linked_list_free(linked_list_t *list, void (*free_value)(void *value));
+
+/**
+ * @brief adds an element to the linked list
+ * 
+ * Add a new element to the end of this linked list. Does nothing if
+ * "list" is NULL.
+ *
+ * @param list list to add the element to
+ * @param value new value to add to the linked list
+ */
+void linked_list_add(linked_list_t *list, void *value);
+
+/**
+ * @brief inserts a value into the linked list
+ *
+ * Inserts a value into the linked list at the specified index. Does
+ * nothing if "list" is NULL.
+ *
+ * @param list list to insert the value into
+ * @param index index where the value should be inserted at (can be equal to list size)
+ * @param value value to insert into the linked list
+ */
+void linked_list_insert(linked_list_t *list, size_t index, void *value);
+
+/**
+ * @brief Get the value in the linked list at the provided index.
+ *
+ * Return the value vstored in the linked list at the provided index. Does
+ * nothing if "list" is NULL.
+ *
+ * @param list list to get the value from
+ * @param index index of the value to return
+ * @return value in the list at the provided index
+ */
+void *linked_list_get(linked_list_t *list, size_t index);
+
+/**
+ * @brief Get the first value that is matched by the provided function
+ *
+ * Return the first value that the function matches. The matches function
+ * parameters are value from the linked list and the provided state.
+ *
+ * @param list list to get the value from
+ * @param matches function to test the values with
+ * @param state state to pass to the matches function
+ * @return first value that matches otherwise NULL
+ */
+void *linked_list_get_first_matching(linked_list_t *list, bool (*matches)(void *item, void *state), void *state);
```

Comment:
Nit: Can we rename state to usrptr or something like that? Or userdata. In general it's the naming used in many other libs.

Reply:
Renamed.